label_studio/io_storages/proxy_api.py

```diff
@@ -0,0 +1,327 @@
+            except (IndexError, ValueError) as e:
+                logger.warning(f"Invalid range header: {rng}: {e}")
+                start = 0
+                end = ''
+
+            """
+            Pass this range as is to storage:
+              - 'bytes=0-'  most likely, browser is requesting just headers
+              - 'bytes=0-0'  most likely, browser is requesting just headers
+            Limit stream to MAX_RANGE bytes:
+              - 'bytes=123456-'  browser is requesting from 123456 to the end of the file
+              - 'bytes=123456-0'  browser is requesting from 123456 to the end of the file
+              - 'bytes=123456-789012'  browser is requesting from 123456 to 789012
+            Not supported:
+              - 'bytes=-1024'  browser is requesting last 1024 bytes - we don't support this
+            """
+            # 'bytes=0-' + 'bytes=0-0'
+            if start == 0 and (end == '' or end == 0):
+                pass
+            # 'bytes=123456-' + 'bytes=123456-0'
+            elif start > 0 and (end == '' or end == 0):
+                end = start + max_range_size
+            # 'bytes=123456-' + 'bytes=123456-789012'
+            elif start >= 0 and end > 0:
+                end = start + max_range_size if end >= start + max_range_size else end
+            # 'bytes=-1024'
+            elif start < 0:
+                logger.warning(f"Start range is negative and not supported: {rng}")
+                start = 0
+                end = start + max_range_size
+            else:
+                logger.warning(f"Range is not covered by logic: {rng}")
+                start = 0
+                end = ''
+
+            range_header = f"bytes={start}-{end}"
+            logger.debug(f">> stream > start: {int(start)/1024/1024} MB")
+            logger.debug(f">> stream > end: {int(end or 0)/1024/1024} MB")
+            logger.debug(f">> stream > range_header: {range_header}")
+            
+        return range_header
+
+    def proxy_data_from_storage(self, request, uri, project, storage):
+        """
+        Proxy the data using iter_chunks directly from storage streaming object.
+        
+        This implementation forwards Range headers to S3 and streams the response
+        directly using StreamingHttpResponse. It avoids any intermediate buffering
+        but doesn't support backward seeking.
+        """
+        try:
+            # Process and limit the range header for downloaded files
+            range_header = self.override_range_header(request)
+            
+            # Use the storage-specific method to get data stream and content type
+            stream, content_type, metadata = storage.get_direct_stream(uri, range_header=range_header)
+            
+            if stream is None:
+                logger.error(f'Failed to get direct stream from storage {storage}')
+                return Response(status=status.HTTP_500_INTERNAL_SERVER_ERROR)
```

Comment:
I know we do this in a few places, but it seems a bit iffy to me whether we should be intentionally returning 500s from the API; usually these indicate errors that we did not anticipate. How would https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/424 sound to you?

Reply:
fixed